source/opt/freeze_spec_constant_value_pass.cpp

```diff
@@ -13,12 +13,23 @@
 // limitations under the License.
 
 #include "source/opt/freeze_spec_constant_value_pass.h"
+
 #include "source/opt/ir_context.h"
+#include "source/opt/set_spec_constant_default_value_pass.h"
+#include "source/util/parse_number.h"
 
 namespace spvtools {
 namespace opt {
 
 Pass::Status FreezeSpecConstantValuePass::Process() {
+  if (spec_ids_.empty()) {
```

Comment:
I don't like that we have two separate paths for doing essentially the same thing, and they do not share any code.  

Could you add a function call `ShouldFixSpecId` that would return true if the given Id is suppose to be fixed.  Then in the orginal code, this function can be called before the switch.

To look for the SpecId Decoration, you can use the decoration manager's member function `ForEachDecoration`.  This should simplify your code because it takes care of looking through decoration groups for you.